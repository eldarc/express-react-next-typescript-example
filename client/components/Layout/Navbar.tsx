import * as React from 'react'
import Link from 'next/link'
import Router from 'next/router'
import { inject, observer } from '@/node_modules/mobx-react'
import { Button, Navbar, NavbarBrand, Nav, NavItem, NavLink, Dropdown, DropdownToggle, DropdownMenu, DropdownItem } from 'reactstrap'

// ~~~ Assets ~~~
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCog, faTachometerAlt } from '@fortawesome/fontawesome-free-solid'
import { IconProp } from '@/node_modules/@fortawesome/fontawesome-svg-core'

// ~~~ Interfaces ~~~
interface IProps {
  [x: string]: any
}

interface IState {
  dropdownOpen: boolean
}

// ~~~ Component ~~~
@inject('store')
@observer
class AppNavbar extends React.Component<IProps, IState> {
  constructor(props: IProps) {
    super(props)

    this.state = {
      dropdownOpen: false
    }
  }

  public toggle = () => {
    this.setState({
      dropdownOpen: !this.state.dropdownOpen
    })
  }

  public signout = async () => {
    const userId = this.props.store.user.userData._id
    await this.props.store.auth.doSignout()
    await this.props.store.overview.unignoreUser(userId)
    Router.push('/overview')
  }

  public render() {
    const userMenu = () => (
      <Nav className="ml-auto d-flex align-items-center" navbar>
        <span className="mr-3 text-gray-500">
          <span className="text-white">{this.props.store.user.userData.username}</span> | Likes: <strong className="text-white">{this.props.store.user.userData.numberOfLikes}</strong>
        </span>
        <Dropdown nav isOpen={this.state.dropdownOpen} toggle={this.toggle}>
          <DropdownToggle nav>
            <FontAwesomeIcon size="lg" icon={faCog as IconProp} />
          </DropdownToggle>
          <DropdownMenu right>
            <Link href="/change-password" passHref>
              <DropdownItem>
                Change password
              </DropdownItem>
            </Link>
            <DropdownItem onClick={this.signout}>Sign out</DropdownItem>
          </DropdownMenu>
        </Dropdown>
      </Nav>
    )

    const guestMenu = () => (
      <div className="ml-auto d-flex align-items-center">
        <Link href="/signin" passHref>
          <Button color="info">Sign in</Button>
        </Link>
      </div>
    )

    return (
      <header>
        <Navbar color="primary" dark expand="md">
          <div className="container">
            <Link href="/" passHref>
              <NavbarBrand>
                LikeBase
              </NavbarBrand>
            </Link>
            <Nav className="mr-auto" navbar>
              <NavItem>
                <Link href="/overview" passHref>
                  <NavLink>
                      <span>
                        <FontAwesomeIcon icon={faTachometerAlt as IconProp} /> Overview
                      </span>
                  </NavLink>
                </Link>
              </NavItem>
            </Nav>
            {this.props.store.auth.isAuthenticated ? userMenu() : guestMenu()}
          </div>
        </Navbar>

        {/*language=SCSS*/}
        <style jsx>{`
        .logo {
          max-height: 30px;
        }
      `}</style>
      </header>
    )
  }
}

export default AppNavbar
