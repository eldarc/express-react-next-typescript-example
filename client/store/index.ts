import { useStaticRendering } from 'mobx-react'
import AuthStore from './modules/AuthStore'
import UserStore from './modules/UserStore'
import OverviewStore from './modules/OverviewStore'
import Cookies from 'js-cookie'

// For SSR static rendering is required.
const isServer = (typeof window === 'undefined')
useStaticRendering(isServer)

// Main root store.
class Store {
  // @ts-ignore (AuthStore is definitely set in the `else` branch in the constructor)
  public auth: AuthStore
  public user: UserStore
  public overview: OverviewStore

  constructor(checkAuth: boolean) {
    this.overview = new OverviewStore()
    this.user = new UserStore()

    // Check if the user is already logged in, and create the store according to that status.
    const token = Cookies.get('token')
    if (checkAuth && token) {
      this.auth = new AuthStore(token)

      // Fetch user data.
      this.user.fetchUserData().then(() => {
        this.overview.ignoreUser(this.user.userData._id)
      })
    } else {
      this.auth = new AuthStore()
    }
  }
}

// Export Store and store initialization function.
export { Store }
let store: Store
export function initializeStore() {
  // Always make a new store if server, otherwise state is shared between requests
  if (isServer) {
    // return new Store(initialData)
    return new Store(false)
  }

  if (!store) {
    store = new Store(true)
  }

  return store
}
