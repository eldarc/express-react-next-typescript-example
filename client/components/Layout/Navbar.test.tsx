/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'
import { Store } from '@/store'

import Navbar from './Navbar'

describe('Navbar', () => {
  it('renders correctly', () => {
    const tree = renderer
        .create(<Navbar store={new Store(false)}/>)
        .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
