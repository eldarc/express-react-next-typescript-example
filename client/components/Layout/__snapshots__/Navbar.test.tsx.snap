// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders correctly 1`] = `
<header
  className="jsx-3485818341"
>
  <nav
    className="navbar navbar-expand-md navbar-dark bg-primary"
  >
    <div
      className="jsx-3485818341 container"
    >
      <a
        className="navbar-brand"
        href="/"
        onClick={[Function]}
      >
        LikeBase
      </a>
      <ul
        className="mr-auto navbar-nav"
      >
        <li
          className="nav-item"
        >
          <a
            className="nav-link"
            href="/overview"
            onClick={[Function]}
          >
            <span
              className="jsx-3485818341"
            >
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-tachometer-alt fa-w-18 "
                data-icon="tachometer-alt"
                data-prefix="fas"
                focusable="false"
                role="img"
                style={Object {}}
                viewBox="0 0 576 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M75.694 480a48.02 48.02 0 0 1-42.448-25.571C12.023 414.3 0 368.556 0 320 0 160.942 128.942 32 288 32s288 128.942 288 288c0 48.556-12.023 94.3-33.246 134.429A48.018 48.018 0 0 1 500.306 480H75.694zM512 288c-17.673 0-32 14.327-32 32 0 17.673 14.327 32 32 32s32-14.327 32-32c0-17.673-14.327-32-32-32zM288 128c17.673 0 32-14.327 32-32 0-17.673-14.327-32-32-32s-32 14.327-32 32c0 17.673 14.327 32 32 32zM64 288c-17.673 0-32 14.327-32 32 0 17.673 14.327 32 32 32s32-14.327 32-32c0-17.673-14.327-32-32-32zm65.608-158.392c-17.673 0-32 14.327-32 32 0 17.673 14.327 32 32 32s32-14.327 32-32c0-17.673-14.327-32-32-32zm316.784 0c-17.673 0-32 14.327-32 32 0 17.673 14.327 32 32 32s32-14.327 32-32c0-17.673-14.327-32-32-32zm-87.078 31.534c-12.627-4.04-26.133 2.92-30.173 15.544l-45.923 143.511C250.108 322.645 224 350.264 224 384c0 35.346 28.654 64 64 64 35.346 0 64-28.654 64-64 0-19.773-8.971-37.447-23.061-49.187l45.919-143.498c4.039-12.625-2.92-26.133-15.544-30.173z"
                  fill="currentColor"
                  style={Object {}}
                />
              </svg>
               Overview
            </span>
          </a>
        </li>
      </ul>
      <div
        className="ml-auto d-flex align-items-center"
      >
        <a
          aria-label={null}
          className="btn btn-info"
          href="/signin"
          onClick={[Function]}
        >
          Sign in
        </a>
      </div>
    </div>
  </nav>
</header>
`;
